node/lib/util/checkout.js

```diff
@@ -601,7 +601,11 @@ function noSuchFileMessage(subName, path) {
 exports.checkoutFiles = co.wrap(function*(repo, options) {
     assert.instanceOf(repo, NodeGit.Repository);
     const resolvedPaths = options.resolvedPaths;
-    const subNames = Object.keys(resolvedPaths);
+
+    // Exception is thrown if we try to get repo info from unopened submodules
```

Comment:
Shouldn't we open the submodules instead?  Or, if the path is the entire submodule, I think we should be able to check out without opening the submodule.

Reply:
With the example I gave above `gm checkout -- .`, opening submodules doesn't sound like the right thing to do because it would result in every submodule being opened. I'm not sure how we can do a checkout on files without opening the submodule, can you explain what that would look like?
`checkout` with files as an argument is a working tree operation and opening a submodule to modify the working tree seems like the wrong thing to do; opening submodules as a side-effect of running `checkout` doesn't sound desirable.

Reply:
We can checkout without opening the submodule by just changing where the submodule pointer in the meta repo's index points.  You can actually see this using regular git.  Just run `git checkout $rev -- foo` on some closed submodule.  I should note that I am not 100% sure that `git meta open` respects this, but if it doesn't, that's a bug.  `git meta commit` definitely should

Reply:
Ah ok I see your point about modifying the submodule pointer. Covering the use cases you described is going to take me more time which I won't have for a while.
What do you think about special casing this behaviour so that we only ignore unopened submodules if `$rev` is empty or `HEAD`? In these cases, ignore the unopened submodules is ok because they are already pointing to the correct commit.
With this approach, we can at least improve the UX for certain use cases now and I could address the other use cases in a future hack day.

Reply:
That seems fine for now.  I guess it's technically wrong in the case where someone has already used regular git checkout to change the submodule pointers to something else, but that's pretty rare.

Reply:
Ok great, I'll work on updating the pr.